### PR TITLE
Add parser for case-iterable, string raw-representable values

### PR DESCRIPTION
**Pull request by stephencelis**
Backporting and updating this from another branch.

### CODE REVIEW

File: Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    Input.SubSequence == Input,
+    Input.Element == UTF8.CodeUnit
+  {
+    .init(toPrefix: { $0.utf8 }, areEquivalent: ==)
+  }
+}
+
+extension Parsers {
+  public struct CaseIterableRawRepresentableParser<
+    Input: Collection, Output: CaseIterable & RawRepresentable, Prefix: Collection
```
Comment (stephencelis) [inline Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:68]: I've been able to relax some constraints by adding a separate generic for the prefix type that is used to test the `starts(with:)` logic in `parse`. So the type is a lil more intense than in the printer branch...

**Comment by mbrandonw** [comment]
I just added a [commit](https://github.com/pointfreeco/swift-parsing/pull/176/commits/69a1eb856edf2e8ab76efe7efb948364402e6482) that brings over the `CaseIterable.md` article from the printer branch, and I realized we could make this stuff work when the raw value is an integer too:

```swift
enum Person: Int, CaseIterable {
  case blob = 4
  case blobJr = 42
}

let peopleParser = Many {
  Person.parser()
} separator: {
  ",".utf8
} terminator: {
  End()
}

var input = "4,42"[...].utf8
XCTAssertEqual(try peopleParser.parse(&input), [.blob, .blobJr])
```

File: Sources/Parsing/Documentation.docc/Articles/Parsers/CaseIterable.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  ","
+  Role.parser()
+}
+.parse("123,member") // (123, .member)
+```
+
+This also works with raw enumerations that are backed by integers:
+
+```swift
+enum Role: String, CaseIterable {
```
Comment (tgrapperon) [inline Sources/Parsing/Documentation.docc/Articles/Parsers/CaseIterable.md:29]:
```suggestion
enum Role: Int, CaseIterable {
```

Comment (mbrandonw) [reply Sources/Parsing/Documentation.docc/Articles/Parsers/CaseIterable.md:29]: Oops thanks!

File: Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift [hunk 3]
```diff
@@ -0,0 +1,133 @@
+extension CaseIterable where Self: RawRepresentable, RawValue == Int {
```
Comment (iampatbrown) [inline Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:1]: @mbrandonw would `RawValue: LosslessStringConvertible & Comparable` also work here?

Comment (tgrapperon) [reply Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:1]: @iampatbrown If I'm not mistaken, the restriction can theoretically be narrowed to being expressible by an integer or floating-point literal, because that's ultimately what will be allowed for `RawRepresentable` enums (other than those with `RawValue` being `ExpressibleByStringLiteral)`. So we can cover every situation with 3 overloads (`ExpressibleBy(Integer/Float/String)Literal`), but I don't know if we can easily extract parser/printers from it.

Being `Comparable` allows to reach `UInt`s for example, but it doesn't help with custom made types. For example, this compiles:
```swift
struct Value: RawRepresentable, Equatable {
  var rawValue: Int
  init(rawValue: Int) { self.rawValue = rawValue }
}

extension Value: ExpressibleByIntegerLiteral {
  init(integerLiteral value: Int) {
    self.rawValue = value
  }
}

enum Enum: Value, CaseIterable {
  case first = 1
}
```

`String` and `Int` are already covering the vast majority of the cases (!) though, but I 'm afraid that `LosslessStringConvertible & Compable` will only partially improve the situation (the `Enum` above would pass through).

File: Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    let areEquivalent: (Input.Element, Input.Element) -> Bool
+
+    @usableFromInline
+    init(
+      toPrefix: @escaping (Output.RawValue) -> Prefix,
+      areEquivalent: @escaping (Input.Element, Input.Element) -> Bool
+    ) {
+      self.toPrefix = toPrefix
+      self.areEquivalent = areEquivalent
+      self.cases = Output.allCases.sorted(by: { $0.rawValue > $1.rawValue })
```
Comment (tgrapperon) [inline Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:118]: Wouldn't it be sufficient to sort by length instead? This would allow dropping the `Comparable` requirement, isn't it?

File: Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift (hunk 3 above)
Comment (iampatbrown) [reply Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:1]: @tgrapperon if I'm not mistaken, the reason `RawValue == Int` works is because `Int` conforms to `LosslessStringConvertible & Comparable`.  The `String.init` being used in `toPrefix: { String($0) }` is `init<T>(_ value: T) where T : LosslessStringConvertible`. 

`RawValue: LosslessStringConvertible & Comparable` should work for `FixedWidthInteger`, `Double` and `Substring`. As for printing, I'm pretty sure that would remain the same using `toPrefix`:

```swift
input.prepend(contentsOf: self.toPrefix(output.rawValue))
```

Maybe I'm missing something though :P

File: Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift (hunk 4 above)
Comment (iampatbrown) [reply Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:118]: @tgrapperon as in something like this?

```swift
self.cases = Output.allCases.sorted(by: {
  toPrefix($0.rawValue).endIndex > toPrefix($1.rawValue).endIndex
})
```

File: Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift (hunk 3 above)
Comment (tgrapperon) [reply Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:1]: @iampatbrown You're right about `LosslessStringConvertible`. I was only reasoning at the `RawRepresentable` enum level, without considering how we can extract parsers/printers from the `RawValue` later.

You're probably right for the constraint, and the example I've provided is quite exotic and a dedicated parser can always be implemented if it ever happens in the wild. Furthermore, the new `RawValue` types that can be reached are all comparable, so it directly works with respect to the comment I made about sorting below.

I didn't check if the compiler was working the same with `Int`/`String` or `LosslessStringConvertible&Comparable`, but if it is, it looks like a better alternative.

File: Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift (hunk 4 above)
Comment (iampatbrown) [reply Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:118]: I guess it doesn't really matter if they have the same length. The main issue would be if a shorter value is ordered before a longer one. This probably means that comparing `.endIndex` would probably work for `Double` but comparing `.rawValue` might not

Comment (iampatbrown) [reply Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:118]: @tgrapperon actually the current implementation would fail for negative `Int`s too. Maybe using `endIndex` is actually the way to go. You can change the current test to negative values to test it out. 

```swift
enum Person: Int, CaseIterable {
  case blob = -4
  case blobJr = -42
}

var input = "-4,-42"[...].utf8
XCTAssertEqual(try peopleParser.parse(&input), [.blob, .blobJr]) // XCTAssertEqual failed
```

Comment (tgrapperon) [reply Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:118]: @iampatbrown With the current state of the PR, this is not really a concern. Sorting is useless but mostly harmless for numeric `RawValue`s, and the other types are collections where sorting by `count` would be equivalent to bin the `lexical` sort.
So it's working indeed, but it performs a little more work than needed. The `Comparable` requirement is fortuitously fulfilled by all supported `RawValue` types. Being `LosslessStringConvertible` automatically provides a canonical way to sort values by projecting them to `.description`, so it can maybe be possible to drop the `Comparable` requirement for a simple `LosslessStringConvertible`, and perform the sort via `.description` only when needed if we manage to exclude numbers (maybe by checking if they're `IntegerLiteralType` or `FloatLiteralType`, or using explicit types directly).

Comment (tgrapperon) [reply Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:118]: > actually the current implementation would fail for negative Ints too.
Good catch!
As I'm getting it, it would work using `.description`, no? I didn't have the time to check out the branch yet.

Comment (tgrapperon) [reply Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:118]: @iampatbrown Ok, I've just checked, and reverse sorting lexically by `.description` seems to do the trick too, but the `toPrefix` seems more efficient/elegant. It doesn't seem to cause issues with same-length strings because they're always different numbers(e.g., "-4" vs "44").

Comment (iampatbrown) [reply Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:118]: Well the sort happens inside the init of the parser where `RawValue` has no constraints apart from being comparable. The `LosslessStringConvertible` was only for a static initilizer. I think you're right about using length though. The sort is to ensure the parser tries to match longer cases first. For `RawValue == String` using `Comparable` works fine.

Comment (tgrapperon) [reply Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift:118]: @iampatbrown That's a chance you did spot the negative `Int` case. This is a very common situation!

**Comment by iampatbrown** [comment]
@stephencelis @mbrandonw Here is a summary of the possible changes from the discussion @tgrapperon and I had:

1. Sort by `endIndex` instead of `rawValue` and drop `Comparable` requirement. `Comparable` works for `String` but not for `Int` because negative numbers will not get sorted by descending length. 

```swift
self.cases = Output.allCases
  .sorted(by: { toPrefix($0.rawValue).endIndex > toPrefix($1.rawValue).endIndex })
```

2. `RawValue: LosslessStringConvertible` could be used in place of `RawValue == Int`. It would add support for additional types and would work with the current  `RawValue == Int` implementation (if sorted by length)

I've included the changes in the below branch if needed.


https://github.com/pointfreeco/swift-parsing/compare/case-iterable-raw-representable...iampatbrown:case-iterable-raw-representable

**Comment by tgrapperon** [comment]
@iampatbrown Since `StringProtocol: LosslessStringConvertible`, aren't we be able to drop the `RawValue: String` overload too?

**Comment by iampatbrown** [comment]
@tgrapperon Yep. It could be dropped. The main reason I didn't suggest it was the redundant call to `String.init` in `toPrefix: { String($0) }`. That's probably insignificant though.

**Comment by tgrapperon** [comment]
Wouldn't `toPrefix: { $0 is String ? $0 : String($0) }` work? (but I feel we're entering the bikeshedding zone).

**Comment by iampatbrown** [comment]
Using `toPrefix: { String($0) }` would be fine and would probably be the same as `toPrefix: { $0 }` in this case.

**Comment by iampatbrown** [comment]
@tgrapperon for some additional context where is the implementation from [swift/String.swift](https://github.com/apple/swift/blob/ba149d964ff49ec9f72090ea9a5f7ded35abdfa2/stdlib/public/core/String.swift#L896-L911)

```swift
  ...
  /// Creates an instance from the description of a given
  /// `LosslessStringConvertible` instance.
  @inlinable @inline(__always)
  public init<T: LosslessStringConvertible>(_ value: T) {
    self = value.description
  }
}

extension String: CustomStringConvertible {
  /// The value of this string.
  ///
  /// Using this property directly is discouraged. Instead, use simple
  /// assignment to create a new constant or variable equal to this string.
  @inlinable
  public var description: String { return self }
}
```

I'd say using the lossless initializer for `String` would pretty much be the same. So I don't see anything wrong with dropping `RawValue == String` as well.

**Comment by tgrapperon** [comment]
From [String.swift:L899](https://github.com/apple/swift/blob/ba149d964ff49ec9f72090ea9a5f7ded35abdfa2/stdlib/public/core/String.swift#L899) and below, it is as optimal as it can be

**Comment by iampatbrown** [comment]
I'll add an edit to the summary.

**Comment by tgrapperon** [comment]
As `value.description` is hit anyway, I wonder if we couldn't use this directly instead of going through `toPrefix` to sort the cases. 
Are there representations where the ordering could be different, that is where `a.description.endIndex < b.description.endIndex` doesn't implies that `toPrefix(String(a)).endIndex < `toPrefix(String(b)).endIndex`?
This is mostly by curiosity, and going through `toPrefix` is definitely a safe and acceptable solution.

I also wonder up to which point comparing indices from different `Strings` gives a meaningful result. I vaguely remember someone from the core team pointing this, but I can't figure out where yet, nor if it ever happened. It would make sense though.
In any case, we can always use the distance from `.startIndex` (that is `.count`), but the complexity is O(n) for `String` (but we're talking about enums' `RawValue`s, not War and Peace).

**Comment by iampatbrown** [comment]
> As value.description is hit anyway, I wonder if we couldn't use this directly instead of going through toPrefix to sort the cases.

I'm not sure how sort could access `.description` unless some type info was passed to the parser init. That being said. I don't think  using `toPrefix($0.rawValue).endIndex` is 100% either. Using `Array(toPrefix($0.rawValue)).count` is probably safer. We could deliberately break `.endIndex` if we really wanted to with a custom `toPrefix`, however, the parser init is not public so right now that shouldn't happen.

**Comment by tgrapperon** [comment]
If `RawValue` is `LosslessStringConvertible`, it has a `.description` that generates a `String` we can measure for sorting. But the more I think about it, and given the implementation from `LosslessStringConvertible`'s `String.init`, I don't think it makes any difference in terms of performance, as all `toPrefix` implementations are trivial. It is probably safer like you proposed, through `toPrefix`, as we then "sort like we parse". I don't have a counterexample where this would be required instead of sorting the `rawValues`'s descriptions, but it looks more resilient like this in any case.

The same argument could be used about `count`. From what I can see, `String.Index` comparison is only meaningful in the same `String`. They internally sort using some integer called `transcodedIndex`, which is stored along another integer called `encodedIndex`. I don't really grasp the correspondence with the internal representation, but I guess it's safer to assume that `transcodedIndex` is string-dependent (I would still like to find a reference to this). In this specific use case, I can't see how this could lead to two `rawValue`s having the same prefix and a wrong ordering, but it's most likely a bad practice in the general case. For this reason, I think it's worth the small expense of sorting using `.count`. The current implementation of the PR is using `Comparable` which is not better in terms of complexity.

**Comment by iampatbrown** [comment]
> If RawValue is LosslessStringConvertible, it has a .description that generates a String we can measure for sorting.

Just want to clarify, right now the sorting occurs inside the init of `CaseIterableRawRepresentableParser` where there is no `LosslessStringConvertible` constraint. That's the reason for using `toPrefix` and why I keep saying `.description` can't be used (without making changes to the init). 

> From what I can see, String.Index comparison is only meaningful in the same String

I'm not exactly sure of the inner workings of `String.Index` but the following from [StringIndex.swift](https://github.com/apple/swift/blob/ba149d964ff49ec9f72090ea9a5f7ded35abdfa2/stdlib/public/core/StringIndex.swift#L282-L287) makes me think it would work:
```swift
extension String.Index {
  @inlinable @inline(__always)
  internal var orderingValue: UInt64 { return _rawBits &>> 14 }
}

extension String.Index: Comparable {
  @inlinable @inline(__always)
  public static func < (lhs: String.Index, rhs: String.Index) -> Bool {
    return lhs.orderingValue < rhs.orderingValue
  }
}
```

I don't think there's any issue using count though. The actual reason I used `endIndex` in the first place was because I thought `Prefix` was a `Sequence` not a `Collection`. I don't think it matters if it's O(n), I'm pretty sure `toPrefix` is going to be O(n) for `LosslessStringConvertible` anyway (apart from when `RawValue == String`)

**Comment by tgrapperon** [comment]
> Just want to clarify, right now the sorting occurs inside the init of CaseIterableRawRepresentableParser where there is no LosslessStringConvertible constraint.

Ah, sorry, I was thinking this internal parser could be constrained too without creating restrictions, as I supposed that all routes creating it are already constrained by `LosslessStringConvertible` (`String` variants were already seen as such in my mind). But this is not an issue, as sorting using simply`.description` is a bad idea (see below).

> I'm not exactly sure of the inner workings of String.Index but the following from [StringIndex.swift](https://github.com/apple/swift/blob/ba149d964ff49ec9f72090ea9a5f7ded35abdfa2/stdlib/public/core/StringIndex.swift#L282-L287) makes me think it would work

Yes, it would work in our case, but `"👨‍👨‍👧‍👧".endIndex > "AB".endIndex` shows it's not a reliable indicator to sort `String`s by length. However, I can't form an example with the same prefixes and "wrong" `endIndex`es, so `endIndex` doesn't seem to be problematic for our intent. But as you said, `.count` is not where time will be lost there.

Furthermore, we **need** to go through `toPrefix` to measure the cases:
```swift
"👨‍👨‍👧".count == "👨‍👨‍👧‍👧".count // == 1
"👨‍👨‍👧‍👧".starts(with: "👨‍👨‍👧") // false

"👨‍👨‍👧".utf8.count != "👨‍👨‍👧‍👧".utf8.count // 18 != 25
"👨‍👨‍👧‍👧".utf8.starts(with: "👨‍👨‍👧".utf8) // true
```
If we sort by `.description` which are `String, we will get ambiguous results when sorting because both have the same `count` (albeit different `endIndex`es!), and if we parse utf8 views, we may match 👨‍👨‍👧 instead of 👨‍👨‍👧‍👧. Because cases go through `toPrefix`, they're correctly ordered, so we must indeed "sort like we parse". This example answers one of my questions above.

I'm sorry for the amount of discussion over this very simple matter! It seems fine like you proposed, maybe with `.count` instead of `.endIndex` for the sake of clarity.

**Review by KeithBird — CHANGES_REQUESTED** [review]
Fix some typos.

**Comment by stephencelis** [comment]
@iampatbrown @tgrapperon Thanks for the in-depth conversation!

Brandon and I spoke about this earlier today. We definitely want to ensure negative numbers are appropriately parsed, so thanks for calling that edge case out! I've pushed a test case and fix.

As far as generalizing to `LosslessStringConvertible` goes, we agree that it would be nice, but are worried that we can't do so as simply as proposed because the formatted string of the value doesn't always have a single representation. For integers, this isn't super problematic, but take an enum that is raw-representable by a `Double`: it would need to be parsed _exactly_ as it is formatted:

```swift
enum Code: Double, CaseIterable {
  case a = 5
  ...
}

Code.parser().parse("5.0") // ✅
Code.parser().parse("5")   // ❌
```

Technically this is even a problem for `Int`s if we wanted to be lenient enough to support parsing a leading "+" sign, though this is arguably a lot less problematic:

```swift
enum Code: Int, CaseIterable {
  case a = 5
  ...
}

Code.parser().parse("5")  // ✅
Code.parser().parse("+5") // ❌
```

We think both of the above could be fixed if `Code.parser()` somehow knew to call out to `Double.parser()` and `Int.parser()` under the hood instead, which might be easier now that we unified all floating-point parsers in #177. For now, though, we think this is in good enough shape to merge.

If either of you come up with a way of supporting `Double` (or if you think you have a reasonable solution for generally parsing any `LosslessStringConvertible` values), please start a discussion or PR :smile:

**Review by mbrandonw — APPROVED** [review]

**Comment by iampatbrown** [comment]
> As far as generalizing to `LosslessStringConvertible` goes, we agree that it would be nice, but are worried that we can't do so as simply as proposed because the formatted string of the value doesn't always have a single representation. 

@stephencelis @mbrandonw Good call. I think `String` and `Int` gives decent coverage. Do you think using  `StringProtocol` and `FixedWidthInteger` respectively would work as the constrains or does that have similar issues?

> We think both of the above could be fixed if Code.parser() somehow knew to call out to Double.parser() and Int.parser() under the hood instead

Sounds like that could require passing a custom parser that incorporates Double.parser()/Int.parser() but also checks the different representations before moving on to the next case. Interesting problem....

> we think this is in good enough shape to merge.

Agreed. Nice work!

**Comment by stephencelis** [comment]
> @stephencelis @mbrandonw Good call. I think `String` and `Int` gives decent coverage. Do you think using `StringProtocol` and `FixedWidthInteger` respectively would work as the constrains or does that have similar issues?

Definitely down for those generalizations! Would you want to explore them?
